test/cpu_reduce_util.hpp

```diff
@@ -29,19 +29,196 @@
 #include <half.hpp>
 #include <limits>
 #include <cmath>
+#include <cassert>
 #include <miopen/miopen.h>
 #include <miopen/reduce_common.hpp>
 
 #include <miopen/bfloat16.hpp>
 
 namespace reduce {
 
+template <typename T>
+static inline bool IsNan(T x)
+{
+    // C++ isnan() is used for float and double
+    return (std::isnan(x));
+};
+
+template <>
+inline bool IsNan<half_float::half>(half_float::half x)
+{
+    return (half_float::isnan(x));
+};
+
+template <typename T>
+static inline bool IsFinite(T x)
+{
+    // C++ isfinite() is used for float and double
+    return (std::isfinite(x));
+};
+
+template <>
+inline bool IsFinite<half_float::half>(half_float::half x)
+{
+    return (half_float::isfinite(x));
+};
+
+template <typename T>
+static inline bool float_equal_one(T x)
+{
+    (void)x;
+    static_assert(static_cast<T>(0), "float_equal_one() is not implemented for this data type");
+    return false;
+};
```

Comment:
We already have a `float_equal` function [here](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/include/miopen/float_equal.hpp#L44). You should use that and add an overload for `half`.

Reply:
float_equal is more expensive.  float_equal_one/float_equal_zero are used to check the alpha/beta values for scaling the tensor. The alpha and beta are passed as float constants, which need not be checked by std::isfinite()

Reply:
@fronteer Let me ask again: is performance important here?

Reply:
It seems like performance is not important as it calls through std::function which is pretty slow for numeric computations.